### PR TITLE
feat: Use TimestampNanosecondArray for timestamps in IOx

### LINKED ISSUE #1084: bug: The `time` column in IOx should be returned as a Timestamp rather than an UInt64
__Steps to reproduce:__
Given our favorite SQL telegraph data

```
alamb@ip-10-0-0-124 influxdb_iox % ./target/debug/influxdb_iox database query 26f7e5a4b7be365b_917b97a92e883afc "select * from cpu limit 10";
./target/debug/influxdb_iox database query 26f7e5a4b7be365b_917b97a92e883afc "select * from cpu limit 10";
+------+-------------------+---------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
| cpu  | host              | time                | usage_guest | usage_guest_nice | usage_idle        | usage_iowait | usage_irq | usage_nice | usage_softirq | usage_steal | usage_system       | usage_user         |
+------+-------------------+---------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
| cpu0 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 65.30408773649165 | 0            | 0         | 0          | 0             | 0           | 18.444666002000673 | 16.251246261217506 |
| cpu1 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 84.43113772402216 | 0            | 0         | 0          | 0             | 0           | 3.193612774446795  | 12.37524950097282  |
| cpu2 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 65.96806387199344 | 0            | 0         | 0          | 0             | 0           | 15.469061876247794 | 18.56287425146831  |
| cpu3 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 84.0478564307993  | 0            | 0         | 0          | 0             | 0           | 3.0907278165770684 | 12.861415752863932 |
| cpu4 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 63.21036889281897 | 0            | 0         | 0          | 0             | 0           | 13.758723828377473 | 23.030907278223218 |
| cpu5 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 83.94815553242313 | 0            | 0         | 0          | 0             | 0           | 2.991026919231221  | 13.0608175473346   |
| cpu6 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 70.85828343276965 | 0            | 0         | 0          | 0             | 0           | 12.87425149699077  | 16.26746506987651  |
| cpu7 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 83.9321357287122  | 0            | 0         | 0          | 0             | 0           | 3.093812375243205  | 12.974051896176206 |
| cpu8 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 74.80079681313936 | 0            | 0         | 0          | 0             | 0           | 10.756972111708253 | 14.442231075949556 |
| cpu9 | MacBook-Pro.local | 1617130130000000000 | 0           | 0                | 83.84845463618315 | 0            | 0         | 0          | 0             | 0           | 3.0907278165434624 | 13.060817547316466 |
+------+-------------------+---------------------+-------------+------------------+-------------------+--------------+-----------+------------+---------------+-------------+--------------------+--------------------+
```

__Expected behaviour:__

the `time` column should be treated as a timestamp (shown as values such as ` 2021-03-30 18:48:50`)

__Actual behaviour:__
Today it is returned as an integer (e.g. 1617130130000000000)

It can be cast to timestamps and things work, but the type should be set correctly already:

```
alamb@ip-10-0-0-124 influxdb_iox % ./target/debug/influxdb_iox database query 26f7e5a4b7be365b_917b97a92e883afc "select time, cast (time as timestamp) from cpu limit 10";

+---------------------+-------------------------------------------+
| time                | CAST(time AS Timestamp(Nanosecond, None)) |
+---------------------+-------------------------------------------+
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
| 1617130130000000000 | 2021-03-30 18:48:50                       |
+---------------------+-------------------------------------------+
```

### PR DESCRIPTION
**Pull request by alamb**
Closes  #1084 😅 

# Rationale
Treat `time` as a `TimetsampNanosecondArray` in IOx (rather than `Int64Array`)

# Changes
- [x] Adds support for emitting time columns from the Mutable Buffer as `Timestamp(TimeUnit::Nanosecond)` (#1089)
- [x] Adds support for accepting and emitting a time column from the Read Buffer as `Timestamp(TimeUnit::Nanosecond)` (part of #1104)
- [x] Updates all tests

# Provenance
This PR has code from  #1089 and https://github.com/influxdata/influxdb_iox/pull/1104 from @e-dard . I chose to make a new PR as the histories and discussions on those PRs has gotten unwieldy. Also I chose not to try and preserve the individual commit history, which while being long and sordid, also was causing massive merge conflicts for me while keeping this long lived branch in sync

# Notes
This also required several upstream Arrow fixes. See #1104  for more details

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: mutable_buffer/src/table.rs [hunk 1]
```diff
@@ -361,73 +359,48 @@ impl Table {
...
-                        match v {
-                            None => builder.append_null(),
-                            Some(s) => builder.append_value(s),
-                        }
-                        .context(ArrowError {})?;
-                    }
-
-                    Arc::new(builder.finish()) as ArrayRef
+                    let iter = vals.iter().map(|s| s.as_deref());
+                    let array = StringArray::from_iter(iter);
```
Comment (alamb) [inline mutable_buffer/src/table.rs:365]: typically we have seen the `from_iter` methods outperform the Builder methods as they can optimize away the bounds checking. And it is less code!

File: internal_types/src/schema.rs [hunk 2]
```diff
@@ -9,12 +9,29 @@ use std::{
...
+    SchemaRef as ArrowSchemaRef, TimeUnit,
 };
 
 /// The name of the timestamp column in the InfluxDB datamodel
 pub const TIME_COLUMN_NAME: &str = "time";
 
+/// The Timezone to use for InfluxDB timezone (should be a constant)
+#[allow(non_snake_case)]
+pub fn TIME_DATA_TIMEZONE() -> Option<String> {
+    // TODO: we should use the "UTC" timezone as that is what the
```
Comment (e-dard) [inline internal_types/src/schema.rs:21]: Might be worth filing a ticket.

**Review by e-dard — APPROVED** [review]
Looked through most of this. Looks good.

**Comment by alamb** [comment]
I will wait for @NGA-TRAN 's review before merging this

File: internal_types/src/schema.rs (hunk 2 above)
Comment (alamb) [reply internal_types/src/schema.rs:21]: https://github.com/influxdata/influxdb_iox/issues/1233

File: arrow_deps/src/util.rs [hunk 3]
```diff
@@ -66,8 +69,13 @@ impl IntoExpr for Expr {
 /// Creates expression like:
 /// start <= time && time < end
 pub fn make_range_expr(start: i64, end: i64, time: impl AsRef<str>) -> Expr {
-    let ts_low = lit(start).lt_eq(col(time.as_ref()));
-    let ts_high = col(time.as_ref()).lt(lit(end));
+    // We need to cast the start and end values to timestamps
+    // the equivalent of:
+    let ts_start = ScalarValue::TimestampNanosecond(Some(start));
+    let ts_end = ScalarValue::TimestampNanosecond(Some(end));
```
Comment (NGA-TRAN) [inline arrow_deps/src/util.rs:75]: Yay!

**Review by NGA-TRAN — APPROVED** [review]
Thanks Andrew for supporting this timestamp and changing a lot of related tests

**Pull request merged**